internal/server/data/migrations.go

```diff
@@ -679,3 +681,245 @@ func cleanCrossOrgGroupMemberships() *migrator.Migration {
+				type groupProviderUser struct {
+					IdentityID uid.ID
+					ProviderID uid.ID
+					Groups     models.CommaSeparatedStrings
+				}
+				stmt := `
+					SELECT identity_id, provider_id, groups
+					FROM provider_users
+					WHERE groups != ''
+				`
+				rows, err := tx.Query(stmt)
+				if err != nil {
+					return fmt.Errorf("migrate existing provider groups: %w", err)
+				}
+
+				groupProviderUsers := []groupProviderUser{}
+				for rows.Next() {
+					var groupProviderUser groupProviderUser
+					if err := rows.Scan(&groupProviderUser.IdentityID, &groupProviderUser.ProviderID, &groupProviderUser.Groups); err != nil {
+						return err
+					}
+
+					groupProviderUsers = append(groupProviderUsers, groupProviderUser)
+				}
+
+				if rows.Err() != nil {
+					return err
+				}
+
+				if err := rows.Close(); err != nil {
+					return err
+				}
+
+				for _, groupProviderUser := range groupProviderUsers {
+					var orgID uid.ID
+					if err := tx.QueryRow(`SELECT organization_id FROM providers WHERE id = ?`, groupProviderUser.ProviderID).Scan(&orgID); err != nil {
+						return err
+					}
+
+					for _, name := range groupProviderUser.Groups {
+						// start by, getting the corresponding infra group which already exists
+						var infraGroupID uid.ID
+						err := tx.QueryRow(`SELECT id FROM groups WHERE name = ? AND organization_id = ?`, name, orgID).Scan(&infraGroupID)
+						if err != nil {
+							return err
+						}
+
+						// check if this provider user group exists yet
+						providerGroup := &models.ProviderGroup{}
+						err = tx.QueryRow(`SELECT * FROM provider_groups WHERE organization_id = ? AND name = ? AND provider_id = ?`, orgID, name, groupProviderUser.ProviderID).Scan(&providerGroup)
+						if err != nil {
+							if !strings.Contains(err.Error(), "no rows in result set") {
+								return err
+							}
+							// need to create the provider group
+							createPGStmt := `
+								INSERT INTO provider_groups(organization_id, created_at, updated_at, provider_id, name)
+								VALUES (?, ?, ?, ?, ?)
+							`
+							_, err = tx.Exec(createPGStmt, orgID, time.Now(), time.Now(), groupProviderUser.ProviderID, name)
+							if err != nil {
+								return err
+							}
+						}
+						// link this provider user as a provider group member
+						_, err = tx.Exec(`
+							INSERT INTO provider_groups_provider_users ("provider_id", "provider_group_name", "provider_user_identity_id")
+							VALUES (?, ?, ?)
+							ON CONFLICT DO NOTHING
+							`,
+							groupProviderUser.ProviderID, name, groupProviderUser.IdentityID)
+						if err != nil {
+							return fmt.Errorf("add provider group member: %w", err)
+						}
+
+						// find the relation that was previously created for this user's group relation
+						var existing uid.ID
+						err = tx.QueryRow(`
+							SELECT group_id FROM identities_groups 
+							WHERE identity_id = ? AND group_id = ? AND provider_id IS NULL`,
```

Comment:
ignore deleted?

Reply:
This one didn't need deleted, but some of the other migration queries did when I went back to look at this, thanks